Quantum-Control-Applications-QuAM/Superconducting/calibrations/06_resonator_spectroscopy_vs_amplitude.py

```diff
@@ -31,48 +30,47 @@
 
 import matplotlib.pyplot as plt
 import numpy as np
+import os
 
-from components import QuAM
-from macros import qua_declaration
-
+from quam_components import QuAM
+from macros import qua_declaration, node_save
 
 ###################################################
 #  Load QuAM and open Communication with the QOP  #
 ###################################################
 # Class containing tools to help handling units and conversions.
 u = unit(coerce_to_integer=True)
 # Instantiate the QuAM class from the state file
-machine = QuAM.load("quam")
+machine = QuAM.load(os.path.join('..', 'configuration', 'quam_state'))
 # Generate the OPX and Octave configurations
 config = machine.generate_config()
-octave_config = machine.octave.get_octave_config()
+octave_config = machine.get_octave_config()
 # Open Communication with the QOP
 qmm = machine.connect()
 
 # Get the relevant QuAM components
-rr1 = machine.active_qubits[0].resonator
-rr2 = machine.active_qubits[1].resonator
-prev_amp1 = rr1.operations["readout"].amplitude
-prev_amp2 = rr2.operations["readout"].amplitude
+resonators = [qb.resonator for qb in machine.active_qubits]
+num_resonators = len(resonators)
+prev_amps = [rr.operations["readout"].amplitude for rr in resonators]
 
 ###################
 # The QUA program #
 ###################
 
 n_avg = 100  # The number of averages
 
-rr1.operations["readout"].amplitude = 0.01
-rr2.operations["readout"].amplitude = 0.01
+# Initial readout amplitude for all resonators
+for rr in resonators:
```

Comment:
I wonder whether we should set the amplitude here. What if the user already calibrated it previously?

Reply:
Since the amplitude sweep is performed w.r.t the current amplitude (via a scaling factor), even if a user has calibrated it previously, it is more convenient to:

1. Change the amplitude value to something "central" from within the script to provide a more meaningful range of amplitudes, even if it is not on resonance.
2. Change it from within the script rather than state.json for more efficient exploration, even if previously calibrated.

I think it's fine to keep for these reasons, but also to demonstrate how to change the amplitude of the machine from python, and to open a discussion about sweeping amplitudes using scaling factors.

Reply:
Your points make sense, but I do see it being overwriten as an issue because it's not clear from the script name that this would happen.

For example, let's say someone has calibrated their entire system including readout pulses. They then notice some inconsistent result, so decide to perform this measurement to see a punch-out. The results look as expected, so they continue their measurements, but now suddenly they can't do proper qubit readout anymore. After some debugging they find the reason to be that their readout amplitude has changed in this script.

wdyt? An alternative is to have this commented out by default and only overwrite it when uncommented

Reply:
Or another option is that I was planning to implement `play(amplitude=...)` instead, which would automatically convert it to the right amplitude scale. But I won't have time until next week for this

Reply:
Yep, you make a good point. For now I will make it commented.

Regarding your point about using something like: `play(amplitude=...)` - it is very convenient usability-wise, but it doesn't solve the fact that the `amp()` command used to input the scale factor has an input range of only -2..2, meaning you can _at most_ double your existing amplitude from the config.

So, unless you start with a high amplitude, you will not have access to a broad range of voltages.